returnn/tf/layers/basic.py

```diff
@@ -2695,22 +2693,41 @@ def get_out_data_from_opts(cls, name, sources, axis="T", **kwargs):
     # Just same format.
     return get_concat_sources_data_template(sources, name="%s_output" % name)
 
+  # noinspection PyMethodOverriding
   @classmethod
-  def get_rec_initial_extra_outputs(cls, batch_dim, rec_layer, axis="T", sources=(), **kwargs):
+  def get_rec_initial_extra_outputs(cls, network, batch_dim, rec_layer, axis="T", sources=(), **kwargs):
     """
+    :param returnn.tf.network.TFNetwork network:
     :param tf.Tensor batch_dim:
     :param returnn.tf.layers.rec.RecLayer|LayerBase rec_layer:
     :param str axis:
     :param list[LayerBase] sources:
     :rtype: dict[str,tf.Tensor]
     """
     data = get_concat_sources_data_template(sources)
-    if axis == "T" and data.time_dim_axis is None:
-      # Assume inside RecLayer.
+    if cls._is_recurrent(axis=axis, network=network, data=data):
       assert all(data.shape)
       return {"state": tf.zeros(data.get_batch_shape(batch_dim=batch_dim), dtype=data.dtype)}
     return {}
 
+  @classmethod
+  def _is_recurrent(cls, axis, network, data):
+    """
+    :param str|DimensionTag axis:
+    :param returnn.tf.network.TFNetwork network:
+    :param Data data:
+    :rtype: bool
+    """
+    if not network.is_inside_rec_layer(inside_loop=False):
```

Comment:
`inside_loop=True`, right? When it is moved out, you want to operate on some normal axis.

But also, this is not the check you want to make. The layer should operate the same inside a rec layer, when you do cum_sum over some other unrelated axis.

Reply:
Ah yes, I was confused about `inside_loop`. You are right.

> But also, this is not the check you want to make. The layer should operate the same inside a rec layer, when you do cum_sum over some other unrelated axis.

But in case I do cumsum over some unrelated axis, then `is_recurrent` should always return `False`. Which is what this check returns, so that should be fine, right?
There were some other issues though, is it correct now?

Reply:
> But in case I do cumsum over some unrelated axis, then `is_recurrent` should always return `False`.

No? Why? `is_recurrent` (I guess you mean `is_inside_rec_layer` and related?) returns whether the layer (network) is inside a rec layer. It is not at all about what axis the `CumSumLayer` would operate on. When you call this function `is_inside_rec_layer`, you do not pass any axis or dim tag to it.

Reply:
With "`_is_recurrent`" (sorry, with `_` before) I meant the function I wrote.
Now I moved this to `tf/util/basic`, and renamed it to `is_axis_from_description_recurrent`.
Maybe the name is not ideal still.

This function is simply meant to replace the old, flawed check `axis == T and data.time_dim_axis is None`, i.e. it should simply say whether `CumsumLayer` should operate step-by-step, or on all steps at once.